programs/drift_vaults/src/state/vault_depositor.rs

```diff
@@ -12,13 +58,353 @@ pub struct VaultDepositor {
+
+            self.update_vault_shares(new_vault_shares, vault)?;
+
+            self.last_withdraw_request_shares =
+                self.last_withdraw_request_shares.safe_div(rebase_divisor)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn deposit(
+        self: &mut VaultDepositor,
+        amount: u64,
+        vault_amount: u64,
+        vault: &mut Vault,
+        now: i64,
+    ) -> Result<()> {
+        validate!(
+            !(vault_amount == 0 && vault.total_shares != 0),
+            ErrorCode::InvalidVaultForNewDepositors,
+            "Vault balance should be non-zero for new stakers to enter"
+        )?;
+
+        vault.apply_rebase(vault_amount)?;
+        self.apply_rebase(vault)?;
+
+        let vault_shares_before = self.checked_vault_shares(vault)?;
+        let total_vault_shares_before = vault.total_shares;
+        let user_vault_shares_before = vault.user_shares;
+
+        let n_shares = vault_amount_to_if_shares(amount, vault.total_shares, vault_amount)?;
+
+        // reset cost basis if no shares
+        self.cost_basis = if vault_shares_before == 0 {
+            amount.cast()?
+        } else {
+            self.cost_basis.safe_add(amount.cast()?)?
+        };
+
+        self.increase_vault_shares(n_shares, vault)?;
+
+        vault.total_shares = vault.total_shares.safe_add(n_shares)?;
+
+        vault.user_shares = vault.user_shares.safe_add(n_shares)?;
+
+        let vault_shares_after = self.checked_vault_shares(vault)?;
+        emit!(VaultDepositorRecord {
+            ts: now,
+            vault: vault.pubkey,
+            user_authority: self.authority,
+            action: VaultDepositorAction::Withdraw,
+            amount,
+            spot_market_index: vault.spot_market_index,
+            vault_amount_before: vault_amount,
+            vault_shares_before,
+            user_vault_shares_before,
+            total_vault_shares_before,
+            vault_shares_after,
+            total_vault_shares_after: vault.total_shares,
+            user_vault_shares_after: vault.user_shares,
+        });
+
+        Ok(())
+    }
+
+    pub fn request_withdraw(
+        self: &mut VaultDepositor,
+        n_shares: u128,
+        vault_amount: u64,
+        vault: &mut Vault,
+        now: i64,
+    ) -> Result<u64> {
+        validate!(
+            n_shares > 0,
+            ErrorCode::InvalidVaultWithdrawSize,
+            "Requested n_shares = 0"
+        )?;
+        validate!(
+            self.last_withdraw_request_shares == 0,
+            ErrorCode::VaultWithdrawRequestInProgress,
+            "Vault withdraw request is already in progress"
+        )?;
+
+        self.last_withdraw_request_shares = n_shares;
+        vault.apply_rebase(vault_amount)?;
+        self.apply_rebase(vault)?;
+
+        let vault_shares_before: u128 = self.checked_vault_shares(vault)?;
+        let total_vault_shares_before = vault.total_shares;
+        let user_vault_shares_before = vault.user_shares;
```

Comment:
you need to check that this is greater than n_shares ...

Reply:
theres a validate below for `vault_shares_before >= n_shares` `InsufficientVaultShares`